File: Src/Entities/Entity.cs
using ECScape.Components;

namespace ECScape.Entities;

internal record Entity
{
    public int Id { get; set; }
    private Dictionary<Type, IComponent> Components { get; set; } = [];

    public Entity(int id)
    {
        Id = id;
    }

    public Entity(int id, List<IComponent> components)
    {
        Id = id;
        foreach (var component in components)
        {
            AddComponent(component);
        }
    }

    public void AddComponent<T>(T component) where T : IComponent
    {
        Components[component.GetType()] = component;
    }

    public bool HasComponent<T>() where T : class, IComponent
    {
        return Components.ContainsKey(typeof(T));
    }

    public bool HasComponent(Type type)
    {
        return Components.ContainsKey(type);
    }

    public T? GetComponent<T>() where T : class, IComponent
    {
        if (!Components.TryGetValue(typeof(T), out var component))
        {
            return null;
        }
        return (T)component;
    }

    public T GetRequiredComponent<T>() where T : class, IComponent
    {
        return GetComponent<T>() ?? throw new InvalidOperationException($"Entity {Id} does not have required component of type {typeof(T).Name}");
    }
}

File: Web/src/entities/Entity.js
class Entity {
    constructor(id) {
        this.Id = id;
        this.Components = new Map();
    }

    addComponent(component) {
        this.Components.set(component.constructor, component);
    }

    removeComponent(componentType) {
        this.Components.delete(componentType);
    }

    hasComponent(componentType) {
        return this.Components.has(componentType);
    }

    getComponent(componentType) {
        return this.Components.get(componentType) || null;
    }

    getRequiredComponent(componentType) {
        const component = this.getComponent(componentType);
        if (!component) {
            throw new Error(`Entity ${this.Id} does not have required component of type ${componentType.name}`);
        }
        return component;
    }
}

export default Entity;